### DISCUSSION TITLE
support of different jetson, any goals to support xavier/orin/etc.?

### DISCUSSION THREAD

**Discussion by oak-tree** [Rubberazer/JETGPIO#10] [Q&A] [answered]

**Comment by Rubberazer** [comment, marked as answer]
UPDATE 24-09-2023:
An update on this, I finally managed to get my hands onto one of the new Orin Nanos and so I updated my library, Orin Nano and Orin NX are basically the same machine for GPIO purposes.

No plan at this point, reason being that I don't have the hardware to test it and I don't like to release things completely untested.
So unless I come across some Orin or something by chance it will not happen, this is also unlikely because on my day job I don't play with this kind of hardware.
I basically created this lib because I got the nano for myself and the libraries available out there were not very enticing imo.

**Comment by Calcu-dev** [comment]
I have access to an NVIDIA Xavier AGX Dev Kit and would be happy to help test/contribute towards integration of the Xavier. I can send you my contact information if you'd like help pursuing this @Rubberazer

**Comment by Rubberazer** [comment]
Hi @Calcu-dev

Thanks for that but not sure to be honest, having the hardware at hand is faster and I can support people after release, plus the Xaviers are powerful but legacy machines by now.
Anyways, if you want to drop me a line and explain more about what your general idea is, motivations e.g. why do you need to support the Xaviers? so forth, we can talk about it: rubberazer@outlook.com

**Reply by Calcu-dev** [reply]
Sent you a message. Thanks!

**Reply by tinChe-** [reply]
Hi @Rubberazer,

I would like to use your library on Xavier NX for low-latency detection of edges on input pins. 
Would update the code myself but I'm unable to find documentation for register locations.
Could you help me find this documentation?

Thank you!

**Reply by Rubberazer** [reply]
@tinChe- 
That would be the Xavier technical reference manual, to be able to download it you are going to need to open an account on the Nvidia site, you will find the manual here:
https://developer.nvidia.com/embedded/downloads

**Reply by tinChe-** [reply]
@Rubberazer 
I have the technical manuals from Orin Nano and Xavier but cannot find the logic how to set the defines in jetgpio.h
Would be very helpful if you could explain for one pin on Orin Nano.

**Reply by Rubberazer** [reply]
@tinChe- 
I am going to give you an example for the Xavier AGX that I shared with somebody else a few months back. This is to **manipulate an output**, but is more than enough to show you how to find out the stuff to manipulate the pins in general, after that is trial and error.
You are going to need the pinmux excel file for your particular model (for AGX would be: Jetson_AGX_Xavier_Series_Pinmux_Configuration_Template_v1.6.xlsm), the technical reference manual and also the kernel sources.

Let's say you want to find out the registers for: GPIO17/A54 which is pin22 in the Xavier AGX developer kit 40 pin header. The starting point would be the pinmux excel file for the
Xavier AGX, in there you will find:</p>

GPIO17 | A54 | SOC_GPIO21 |   | GPIO3_PQ.01
-- | -- | -- | -- | --



<p>The next step would be to find that on the TRM.</p>
<p>Pinmux set up:</p>
<p>PADCTL_CAM_SOC_GPIO21_0<br/>
Offset: 0x58<br/>
Read/Write:
RW<br/>
Parity Protection: N<br/>
SCR Protection:
SCR_SOC_GPIO21_0<br/>
Reset: 0x00000454
(0bxxxx,xxxx,xxxx,xxxx,xxx0,x1x0,x1x1,0100)</p>
<p>The base address should be: PADCTL_A0&nbsp; 0x02430000 according
to below:</p>
<p>8.5.4.1 CAM PAD Control Registers<br/>
Note: The CAM PAD Control
Registers are collectively called PADCTL_A0 in the System Address
Map.</p>
<p>and so the Pinmux address for that specific pin should be:
0x02430058, we want an output so we should be writing the following
value into it: 0x0, although the value: 0x40 would be valid as well.</p>
<p>The next step would be to find out how to manipulate the pin
itself, in the TRM we have:</p>
<p>Table 8.46 GPIO's in Different Pad Control Groups the pin is on GPIO_CTL2:<br/>
GPIO_CTL2&nbsp;&nbsp;
UART, EDP, CAM M, N, P, <b>Q</b>, R, X , Y, Z</p>
<p><font face="C059"><font size="2" style="font-size: 10pt">Now our
pin&nbsp; is:&nbsp;PQ.01, and so it belongs to group: Q, bank&nbsp;
2, port 6 and the pin is 1.</font></font></p>
<p><font face="C059"><font size="2" style="font-size: 10pt">As it
belongs to GPIO_CTL2, the address should be: GPIO_CTL0_GPIO2&nbsp;&nbsp;
0x02212000.</font></font></p>
<p>Now, as the following information is not on the TRM we have to
jump over the kernel sources, the file to look out for is:
gpio-tegra186.c, in there we find:</p>
<p>#define TEGRA194_MAIN_GPIO_PORT(_name, _bank, _port, _pins)&nbsp;&nbsp;&nbsp;
\<br/>
&nbsp;&nbsp; &nbsp;[TEGRA194_MAIN_GPIO_PORT_##_name] = {&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; \<br/>
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;
.name = #_name,&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; \<br/>
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;
.bank = _bank,&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; \<br/>
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;
.port = _port,&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; \<br/>
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;
.pins = _pins,&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; \<br/>
&nbsp;&nbsp; &nbsp;}<br/>
<br/>
static
const struct tegra_gpio_port tegra194_main_ports[] = {<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( A, 1, 2, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( B, 4, 7, 2),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( C, 4, 3, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( D, 4, 4, 4),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( E, 4, 5, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( F, 4, 6, 6),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( G, 4, 0, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( H, 4, 1, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( I, 4, 2, 5),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( J, 5, 1, 6),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( K, 3, 0, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( L, 3, 1, 4),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( M, 2, 3, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( N, 2, 4, 3),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( O, 5, 0, 6),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( P, 2, 5, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( Q, 2, 6, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( R, 2, 7, 6),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( S, 3, 3, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( T, 3, 4, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( U, 3, 5, 1),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( V, 1, 0, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( W, 1, 1, 2),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( X, 2, 0, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( Y, 2, 1, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT( Z, 2, 2, 8),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT(FF, 3, 2, 2),<br/>
&nbsp;&nbsp;
&nbsp;TEGRA194_MAIN_GPIO_PORT(GG, 0, 0, 2)<br/>
};</p>
<p>We care about this line:&nbsp; TEGRA194_MAIN_GPIO_PORT( Q, 2, 6,
8) =&gt; port is number 6, each port from the base address is 0x200
and so for 6 we get:</p>
<ul>
	<li><p style="margin-bottom: 0cm">For 0:0x02212000 
	</p>
	<li><p style="margin-bottom: 0cm">For 1:0x02212200 
	</p>
	<li><p style="margin-bottom: 0cm">For 2:0x02212400 
	</p>
	<li><p style="margin-bottom: 0cm">For 3:0x02212600 
	</p>
	<li><p style="margin-bottom: 0cm">For 4:0x02212800 
	</p>
	<li><p style="margin-bottom: 0cm">For 5:0x02212A00 
	</p>
	<li><p>For 6:0x02212C00 
	</p>
</ul>
<p>And so. so far our port address would be: 0x02212C00, now we have
to add the pin, each pin moves in 0x20 from 0 so for Q0.1 =&gt;
0x02212C20</p>
<p>This is the actual starting point, now if we believe the code in
the driver file we have these macros:</p>
<p>/* control registers */<br/>
#define TEGRA186_GPIO_ENABLE_CONFIG
0x00<br/>
#define&nbsp; TEGRA186_GPIO_ENABLE_CONFIG_ENABLE
BIT(0)<br/>
#define&nbsp; TEGRA186_GPIO_ENABLE_CONFIG_OUT
BIT(1)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_TYPE_NONE (0x0 &lt;&lt;
2)<br/>
#define&nbsp; TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_TYPE_LEVEL
(0x1 &lt;&lt; 2)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_TYPE_SINGLE_EDGE (0x2 &lt;&lt;
2)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_TYPE_DOUBLE_EDGE (0x3 &lt;&lt;
2)<br/>
#define&nbsp; TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_TYPE_MASK
(0x3 &lt;&lt; 2)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_TRIGGER_LEVEL BIT(4)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_DEBOUNCE BIT(5)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_INTERRUPT BIT(6)<br/>
#define&nbsp;
TEGRA186_GPIO_ENABLE_CONFIG_TIMESTAMP_FUNC BIT(7)<br/>
<br/>
#define
TEGRA186_GPIO_DEBOUNCE_CONTROL 0x04<br/>
#define&nbsp;
TEGRA186_GPIO_DEBOUNCE_CONTROL_THRESHOLD(x) ((x) &amp; 0xff)<br/>
<br/>
#define
TEGRA186_GPIO_INPUT 0x08<br/>
#define&nbsp; TEGRA186_GPIO_INPUT_HIGH
BIT(0)<br/>
<br/>
#define TEGRA186_GPIO_OUTPUT_CONTROL 0x0c<br/>
#define&nbsp;
TEGRA186_GPIO_OUTPUT_CONTROL_FLOATED BIT(0)<br/>
<br/>
#define
TEGRA186_GPIO_OUTPUT_VALUE 0x10<br/>
#define&nbsp;
TEGRA186_GPIO_OUTPUT_VALUE_HIGH BIT(0)<br/>
<br/>
#define
TEGRA186_GPIO_INTERRUPT_CLEAR 0x14<br/>
<br/>
#define
TEGRA186_GPIO_INTERRUPT_STATUS(x) (0x100 + (x) * 4)</p>
<p>The ones we care about for an output are:</p>
<p>#define TEGRA186_GPIO_ENABLE_CONFIG 0x00 =&gt; this would be our
0x02212C20</p>
<p>#define TEGRA186_GPIO_OUTPUT_CONTROL 0x0c =&gt; this would be:
0x02212C2C</p>
<p>#define TEGRA186_GPIO_OUTPUT_VALUE 0x10 =&gt; this would be:
0x02212C30</p>
</body>
</html>

And finally to test it with devmem as follows:

1.     set up pinmux: sudo busybox devmem 0x02430058 32 0x0
2.     CFG register drive up/down: sudo busybox devmem 0x0243005C 32 0x01f1f000
3.     enable gpio: sudo busybox devmem 0x02212C20 32 0x3
4.     activate output: sudo busybox devmem 0x02212C2C 32 0x0
5.     turn on: sudo busybox devmem 0x02212C30 32 0x1
6.     turn off: sudo busybox devmem 0x02212C30 32 0x0
